index.d.ts

```diff
@@ -335,394 +104,16 @@ declare class PProgress<ValueType> extends Promise<ValueType> {
-		ValueType1,
-		ValueType2,
-		ValueType3,
-		ValueType4,
-		ValueType5,
-		ValueType6
-	>(
-		promises: [
-			PromiseLike<ValueType1> | PProgress.PromiseFactory<ValueType1>,
-			PromiseLike<ValueType2> | PProgress.PromiseFactory<ValueType2>,
-			PromiseLike<ValueType3> | PProgress.PromiseFactory<ValueType3>,
-			PromiseLike<ValueType4> | PProgress.PromiseFactory<ValueType4>,
-			PromiseLike<ValueType5> | PProgress.PromiseFactory<ValueType5>,
-			PromiseLike<ValueType6> | PProgress.PromiseFactory<ValueType6>
-		]
-	): PProgress<
-		[ValueType1, ValueType2, ValueType3, ValueType4, ValueType5, ValueType6]
-	>;
-	static all<ValueType1, ValueType2, ValueType3, ValueType4, ValueType5>(
-		promises: [
-			PromiseLike<ValueType1> | PProgress.PromiseFactory<ValueType1>,
-			PromiseLike<ValueType2> | PProgress.PromiseFactory<ValueType2>,
-			PromiseLike<ValueType3> | PProgress.PromiseFactory<ValueType3>,
-			PromiseLike<ValueType4> | PProgress.PromiseFactory<ValueType4>,
-			PromiseLike<ValueType5> | PProgress.PromiseFactory<ValueType5>
-		]
-	): PProgress<[ValueType1, ValueType2, ValueType3, ValueType4, ValueType5]>;
-	static all<ValueType1, ValueType2, ValueType3, ValueType4>(
-		promises: [
-			PromiseLike<ValueType1> | PProgress.PromiseFactory<ValueType1>,
-			PromiseLike<ValueType2> | PProgress.PromiseFactory<ValueType2>,
-			PromiseLike<ValueType3> | PProgress.PromiseFactory<ValueType3>,
-			PromiseLike<ValueType4> | PProgress.PromiseFactory<ValueType4>
-		]
-	): PProgress<[ValueType1, ValueType2, ValueType3, ValueType4]>;
-	static all<ValueType1, ValueType2, ValueType3>(
-		promises: [
-			PromiseLike<ValueType1> | PProgress.PromiseFactory<ValueType1>,
-			PromiseLike<ValueType2> | PProgress.PromiseFactory<ValueType2>,
-			PromiseLike<ValueType3> | PProgress.PromiseFactory<ValueType3>
-		]
-	): PProgress<[ValueType1, ValueType2, ValueType3]>;
-	static all<ValueType1, ValueType2>(
-		promises: [
-			PromiseLike<ValueType1> | PProgress.PromiseFactory<ValueType1>,
-			PromiseLike<ValueType2> | PProgress.PromiseFactory<ValueType2>
-		]
-	): PProgress<[ValueType1, ValueType2]>;
-	static all<ValueType1>(
-		promises: [PromiseLike<ValueType1> | PProgress.PromiseFactory<ValueType1>]
-	): PProgress<[ValueType1]>;
-	static all<AllValuesType>(
-		promises: Iterable<PProgress.PromiseFactory<AllValuesType>>,
-		options: PProgress.Options
-	): PProgress<Iterable<AllValuesType>>;
-	static all<AllValuesType>(
-		promises: Iterable<
-			PromiseLike<AllValuesType> | PProgress.PromiseFactory<AllValuesType>
-		>
-	): PProgress<Iterable<AllValuesType>>;
+	static all<Promises extends Array<PProgress.PromiseFactory<unknown> | PromiseLike<unknown>>>(
+		promises: readonly [...Promises],
+		options?: PProgress.Options
+	): PProgress<{
+		[Promise_ in keyof Promises]: Promises[Promise_] extends PromiseLike<unknown> ? Awaited<Promises[Promise_]> : (Promises[Promise_] extends PProgress.PromiseFactory<unknown> ? Awaited<ReturnType<Promises[Promise_]>> : Promises[Promise_])
```

Comment:
Personally find this hard to read. I don't think the TS plugin for ESLint has a rule for this but, I might suggest something like this:

```suggestion
		[Promise_ in keyof Promises]: (
			Promises[Promise_] extends PromiseLike<unknown>
				? Awaited<Promises[Promise_]>
				: (
					Promises[Promise_] extends PProgress.PromiseFactory<unknown>
						? Awaited<ReturnType<Promises[Promise_]>>
						: Promises[Promise_]
				)
		)

```